File: vocal_flutter/lib/sheets/food_database_sheet.dart
// Lightweight food-database search — Flutter port of FoodDatabaseSheet.swift.
//
// Lets the user paste a UPC/EAN and pull it through the existing barcode
// resolver (USDA Branded → Open Food Facts on the worker side). This is a
// SCAFFOLD: there's no free-text food-name search endpoint yet, so name
// queries surface a "coming soon" hint and suggest the camera/voice path
// instead. The barcode-lookup path is fully wired.

import 'package:flutter/material.dart';
import 'package:flutter/services.dart';
import 'package:provider/provider.dart';

import '../models/models.dart';
import '../services/barcode_api.dart';
import '../state/app_model.dart';
import '../theme/theme.dart';
import '../widgets/components.dart';

Future<void> showFoodDatabaseSheet(BuildContext context) {
  return showModalBottomSheet(
    context: context,
    isScrollControlled: true,
    backgroundColor: Palette.ink,
    barrierColor: Colors.black.withOpacity(0.6),
    builder: (_) => const FractionallySizedBox(
      heightFactor: 0.85,
      child: FoodDatabaseSheet(),
    ),
  );
}

class FoodDatabaseSheet extends StatefulWidget {
  const FoodDatabaseSheet({super.key});

  @override
  State<FoodDatabaseSheet> createState() => _FoodDatabaseSheetState();
}

class _FoodDatabaseSheetState extends State<FoodDatabaseSheet> {
  final TextEditingController _query = TextEditingController();
  bool _lookingUp = false;
  MealEntry? _lastResult;
  String? _lastError;

  @override
  void dispose() {
    _query.dispose();
    super.dispose();
  }

  String get _trimmed => _query.text.trim();

  /// True when the query is a plausible barcode (8–14 digits after
  /// stripping whitespace). Routes the same input box to the barcode
  /// API without forcing the user to pick a mode.
  bool get _queryIsCode {
    final t = _trimmed;
    if (t.isEmpty) return false;
    final digitsOnly = t.split('').every(
        (c) => c.codeUnitAt(0) >= 0x30 && c.codeUnitAt(0) <= 0x39);
    if (!digitsOnly) return false;
    return t.length >= 8 && t.length <= 14;
  }

  Future<void> _runLookup() async {
    final q = _trimmed;
    if (q.isEmpty) return;
    if (_queryIsCode) {
      await _lookupCode(q);
    } else {
      setState(() {
        _lastError = 'Free-text name search is coming soon. For now, '
            'scan the barcode with the camera or describe it via voice.';
        _lastResult = null;
      });
    }
  }

  Future<void> _lookupCode(String code) async {
    setState(() {
      _lookingUp = true;
      _lastError = null;
    });
    try {
      final result = await BarcodeApi.lookup(code);
      if (!mounted) return;
      final meal = MealEntry(
        name: result.meal.name,
        detail: result.meal.detail,
        calories: result.meal.kcal,
        protein: result.meal.proteinG,
        carbs: result.meal.carbsG,
        fat: result.meal.fatG,
        loggedAt: DateTime.now(),
        slot: MealSlot.fromRaw(result.meal.slot),
        source: MealSource.barcode,
      );
      setState(() {
        _lastResult = meal;
        _lastError = null;
        _lookingUp = false;
      });
    } on BarcodeApiNotFound {
      if (!mounted) return;
      setState(() {
        _lastResult = null;
        _lastError = "Couldn't find that code in the food database.";
        _lookingUp = false;
      });
    } catch (e) {
      if (!mounted) return;
      setState(() {
        _lastResult = null;
        _lastError = 'Lookup failed. $e';
        _lookingUp = false;
      });
    }
  }

  void _logResult() {
    final meal = _lastResult;
    if (meal == null) return;
    context.read<AppModel>().addMeal(meal);
    HapticFeedback.mediumImpact();
    Navigator.of(context).maybePop();
  }

  @override
  Widget build(BuildContext context) {
    return Container(
      color: Palette.ink,
      padding: EdgeInsets.only(
          bottom: MediaQuery.of(context).viewInsets.bottom),
      child: SafeArea(
        child: Padding(
          padding: const EdgeInsets.fromLTRB(24, 18, 24, 24),
          child: Column(
            crossAxisAlignment: CrossAxisAlignment.start,
            children: [
              _header(),
              const SizedBox(height: 16),
              _searchField(),
              const SizedBox(height: 16),
              if (_lookingUp)
                _loadingRow()
              else if (_lastResult != null)
                _resultCard(_lastResult!)
              else if (_lastError != null)
                _errorRow(_lastError!)
              else
                _placeholderHint(),
              const Spacer(),
            ],
          ),
        ),
      ),
    );
  }

  Widget _header() {
    return Row(
      crossAxisAlignment: CrossAxisAlignment.start,
      children: [
        Expanded(
          child: Column(
            crossAxisAlignment: CrossAxisAlignment.start,
            children: [
              const Eyebrow('FOOD DATABASE', color: Palette.pulse),
              const SizedBox(height: 6),
              Text(
                'Search by name or barcode.',
                style: AppType.serif(28, weight: FontWeight.w500),
              ),
            ],
          ),
        ),
        GestureDetector(
          onTap: () => Navigator.of(context).maybePop(),
          child: Container(
            width: 30,
            height: 30,
            decoration: BoxDecoration(
              shape: BoxShape.circle,
              border: Border.all(color: Palette.hairlineStrong),
            ),
            child: const Icon(Icons.close, size: 13, color: Palette.ash),
          ),
        ),
      ],
    );
  }

  Widget _searchField() {
    return Container(
      padding: const EdgeInsets.symmetric(horizontal: 14, vertical: 4),
      decoration: BoxDecoration(
        color: Palette.inkSurface,
        borderRadius: BorderRadius.circular(Radii.sm),
        border: Border.all(color: Palette.hairlineStrong),
      ),
      child: Row(
        children: [
          const Icon(Icons.search, size: 14, color: Palette.smoke),
          const SizedBox(width: 10),
          Expanded(
            child: TextField(
              controller: _query,
              onChanged: (_) => setState(() {}),
              onSubmitted: (_) => _runLookup(),
              style: AppType.body(14, color: Palette.bone),
              keyboardType: TextInputType.text,
              autocorrect: false,
              decoration: InputDecoration(
                isCollapsed: true,
                contentPadding: const EdgeInsets.symmetric(vertical: 12),
                border: InputBorder.none,
                enabledBorder: InputBorder.none,
                focusedBorder: InputBorder.none,
                hintText: '"Greek yogurt" or 0049000028058',
                hintStyle: AppType.body(14, color: Palette.smoke),
              ),
            ),
          ),
          GestureDetector(
            onTap: _trimmed.isEmpty ? null : _runLookup,
            child: Container(
              margin: const EdgeInsets.symmetric(vertical: 6),
              width: 32,
              height: 32,
              decoration: BoxDecoration(
                shape: BoxShape.circle,
                color: _trimmed.isEmpty
                    ? Palette.hairlineStrong
                    : Palette.paper,
              ),
              child: Icon(
                Icons.arrow_forward,
                size: 16,
                color:
                    _trimmed.isEmpty ? Palette.smoke : Palette.ink,
              ),
            ),
          ),
        ],
      ),
    );
  }

  Widget _loadingRow() {
    return Container(
      width: double.infinity,
      padding: const EdgeInsets.all(20),
      decoration: cardDecoration(),
      child: Row(
        children: [
          const SizedBox(
            width: 14,
            height: 14,
            child: CircularProgressIndicator(
                strokeWidth: 2, color: Palette.bone),
          ),
          const SizedBox(width: 10),
          Text(
            'Looking up...',
            style: AppType.body(13, color: Palette.smoke),
          ),
        ],
      ),
    );
  }

  Widget _resultCard(MealEntry meal) {
    return Container(
      width: double.infinity,
      padding: const EdgeInsets.all(20),
      decoration: cardDecoration(),
      child: Column(
        crossAxisAlignment: CrossAxisAlignment.start,
        children: [
          const Eyebrow('RESULT'),
          const SizedBox(height: 12),
          Text(
            meal.name,
            style: AppType.serif(22, weight: FontWeight.w500),
          ),
          const SizedBox(height: 12),
          Row(
            children: [
              Text(
                '${meal.calories} kcal',
                style: AppType.mono(13, weight: FontWeight.w600),
              ),
              const SizedBox(width: 10),
              MacroPill(letter: 'P', value: meal.protein, tint: Palette.protein),
              const SizedBox(width: 6),
              MacroPill(letter: 'C', value: meal.carbs, tint: Palette.carbs),
              const SizedBox(width: 6),
              MacroPill(letter: 'F', value: meal.fat, tint: Palette.fat),
            ],
          ),
          if (meal.detail.isNotEmpty) ...[
            const SizedBox(height: 10),
            Text(
              meal.detail,
              style: AppType.body(12, color: Palette.smoke),
            ),
          ],
          const SizedBox(height: 14),
          Row(
            children: [
              GhostButton(
                title: 'Clear',
                fullWidth: false,
                onTap: () {
                  setState(() {
                    _lastResult = null;
                    _lastError = null;
                    _query.clear();
                  });
                },
              ),
              const SizedBox(width: 12),
              VoltageButton(
                title: 'Log it',
                icon: Icons.check,
                fullWidth: false,
                onTap: _logResult,
              ),
            ],
          ),
        ],
      ),
    );
  }

  Widget _errorRow(String err) {
    return Container(
      width: double.infinity,
      padding: const EdgeInsets.all(20),
      decoration: cardDecoration(),
      child: Text(
        err,
        style: AppType.body(13, color: Palette.pulse),
      ),
    );
  }

  Widget _placeholderHint() {
    return Container(
      width: double.infinity,
      padding: const EdgeInsets.all(20),
      decoration: cardDecoration(),
      child: Column(
        crossAxisAlignment: CrossAxisAlignment.start,
        children: [
          Text(
            'Type to search',
            style: AppType.serif(20, weight: FontWeight.w500),
          ),
          const SizedBox(height: 8),
          Text(
            'Paste a UPC/EAN to look it up in the food database. '
            "Free-text name search is rolling out — for now, the camera "
            'shutter or voice still works for foods without a barcode.',
            style: AppType.body(13, color: Palette.smoke),
          ),
        ],
      ),
    );
  }
}

File: VoCal/VoCal/FoodDatabaseSheet.swift
//
//  FoodDatabaseSheet.swift
//  VoCal
//
//  Lightweight food-database search. Lets the user type a product name
//  or a UPC/EAN code and pull it through the existing barcode resolver —
//  the path that already chains USDA Branded → Open Food Facts and back
//  to the BarcodeAPI client, no new backend.
//
//  This is a SCAFFOLD: there's no proper free-text food-name search
//  endpoint yet, so the experience is intentionally narrow — code lookup
//  works today; name lookup shows a "coming soon" hint and falls back to
//  prompting the user to scan/snap instead. When a real food search ships,
//  it slots in here at the existing search field without touching the
//  surrounding sheet UI.
//
//  Future-vision shape: the row model has placeholder room for a
//  cuisine / region / cooking-method eyebrow so the regional-recipe
//  intelligence work has a visible home when it's wired up.
//

import SwiftUI

struct FoodDatabaseSheet: View {
    @EnvironmentObject private var appModel: AppModel
    @Environment(\.dismiss) private var dismiss

    @State private var query: String = ""
    @State private var lookingUp: Bool = false
    @State private var lastResult: MealEntry?
    @State private var lastError: String?

    private var trimmedQuery: String {
        query.trimmingCharacters(in: .whitespaces)
    }

    /// True when the query is a plausible barcode (8–14 digits after we
    /// strip whitespace). Lets us route the same input box to the
    /// barcode API without forcing the user to pick a mode.
    private var queryIsCode: Bool {
        let digits = trimmedQuery.unicodeScalars
            .filter { CharacterSet.decimalDigits.contains($0) }
            .map { String($0) }
            .joined()
        return digits.count == trimmedQuery.count && (8...14).contains(digits.count)
    }

    var body: some View {
        ZStack {
            Theme.Palette.ink.ignoresSafeArea()
            VStack(alignment: .leading, spacing: 16) {
                header
                searchField
                if lookingUp {
                    loadingRow
                } else if let result = lastResult {
                    resultCard(result)
                } else if let err = lastError {
                    errorRow(err)
                } else {
                    placeholderHint
                }
                Spacer(minLength: 0)
            }
            .padding(.horizontal, 24)
            .padding(.top, 18)
            .padding(.bottom, 24)
        }
    }

    private var header: some View {
        HStack(alignment: .top) {
            VStack(alignment: .leading, spacing: 6) {
                Text("FOOD DATABASE")
                    .eyebrow(Theme.Palette.pulse)
                Text("Search by name or barcode.")
                    .font(Theme.Font.serif(28, weight: .medium))
                    .foregroundStyle(Theme.Palette.bone)
            }
            Spacer()
            Button { dismiss() } label: {
                Image(systemName: "xmark")
                    .font(.system(size: 13, weight: .semibold))
                    .foregroundStyle(Theme.Palette.ash)
                    .frame(width: 30, height: 30)
                    .background(Circle().strokeBorder(Theme.Palette.hairlineStrong, lineWidth: 1))
            }
            .buttonStyle(.plain)
        }
    }

    private var searchField: some View {
        HStack(spacing: 10) {
            Image(systemName: "magnifyingglass")
                .font(.system(size: 13, weight: .semibold))
                .foregroundStyle(Theme.Palette.smoke)
            TextField(
                "",
                text: $query,
                prompt: Text("\"Greek yogurt\" or 0049000028058").foregroundStyle(Theme.Palette.smoke)
            )
            .textInputAutocapitalization(.never)
            .foregroundStyle(Theme.Palette.bone)
            .onSubmit { runLookup() }
            Button {
                runLookup()
            } label: {
                Image(systemName: "arrow.right")
                    .font(.system(size: 13, weight: .bold))
                    .foregroundStyle(Theme.Palette.ink)
                    .frame(width: 32, height: 32)
                    .background(Circle().fill(Theme.Palette.paper))
            }
            .buttonStyle(.plain)
            .disabled(trimmedQuery.isEmpty)
        }
        .padding(.vertical, 12)
        .padding(.horizontal, 14)
        .background(
            RoundedRectangle(cornerRadius: Theme.Radius.sm, style: .continuous)
                .fill(Theme.Palette.inkSurface)
                .overlay(
                    RoundedRectangle(cornerRadius: Theme.Radius.sm, style: .continuous)
                        .strokeBorder(Theme.Palette.hairlineStrong, lineWidth: 1)
                )
        )
    }

    private var loadingRow: some View {
        HStack(spacing: 10) {
            ProgressView().controlSize(.small).tint(Theme.Palette.bone)
            Text("Looking up...")
                .font(.system(size: 13))
                .foregroundStyle(Theme.Palette.smoke)
        }
        .padding(20)
        .frame(maxWidth: .infinity, alignment: .leading)
        .background(
            RoundedRectangle(cornerRadius: Theme.Radius.md, style: .continuous)
                .fill(Theme.Palette.inkSurface)
                .overlay(
                    RoundedRectangle(cornerRadius: Theme.Radius.md, style: .continuous)
                        .strokeBorder(Theme.Palette.hairline, lineWidth: 1)
                )
        )
    }

    private func resultCard(_ meal: MealEntry) -> some View {
        VStack(alignment: .leading, spacing: 12) {
            Text("RESULT")
                .eyebrow()
            Text(meal.name)
                .font(Theme.Font.serif(22, weight: .medium))
                .foregroundStyle(Theme.Palette.bone)
            HStack(spacing: 10) {
                Text("\(meal.calories) kcal")
                    .font(Theme.Font.mono(13, weight: .semibold))
                    .foregroundStyle(Theme.Palette.bone)
                MacroPill(letter: "P", value: meal.protein, tint: Theme.Palette.protein)
                MacroPill(letter: "C", value: meal.carbs, tint: Theme.Palette.carbs)
                MacroPill(letter: "F", value: meal.fat, tint: Theme.Palette.fat)
            }
            if !meal.detail.isEmpty {
                Text(meal.detail)
                    .font(.system(size: 12))
                    .foregroundStyle(Theme.Palette.smoke)
            }
            HStack(spacing: 12) {
                GhostButton(title: "Clear", fullWidth: false) {
                    lastResult = nil
                    lastError = nil
                    query = ""
                }
                VoltageButton(title: "Log it", icon: "checkmark", fullWidth: false) {
                    appModel.addMeal(meal)
                    UIImpactFeedbackGenerator(style: .medium).impactOccurred()
                    dismiss()
                }
            }
        }
        .padding(20)
        .frame(maxWidth: .infinity, alignment: .leading)
        .background(
            RoundedRectangle(cornerRadius: Theme.Radius.md, style: .continuous)
                .fill(Theme.Palette.inkSurface)
                .overlay(
                    RoundedRectangle(cornerRadius: Theme.Radius.md, style: .continuous)
                        .strokeBorder(Theme.Palette.hairline, lineWidth: 1)
                )
        )
    }

    private func errorRow(_ err: String) -> some View {
        Text(err)
            .font(.system(size: 13))
            .foregroundStyle(Theme.Palette.pulse)
            .padding(20)
            .frame(maxWidth: .infinity, alignment: .leading)
            .background(
                RoundedRectangle(cornerRadius: Theme.Radius.md, style: .continuous)
                    .fill(Theme.Palette.inkSurface)
                    .overlay(
                        RoundedRectangle(cornerRadius: Theme.Radius.md, style: .continuous)
                            .strokeBorder(Theme.Palette.hairline, lineWidth: 1)
                    )
            )
    }

    private var placeholderHint: some View {
        VStack(alignment: .leading, spacing: 12) {
            Text("Type to search")
                .font(Theme.Font.serif(20, weight: .medium))
                .foregroundStyle(Theme.Palette.bone)
            Text("Paste a UPC/EAN to look it up in the food database. Free-text name search is rolling out — for now, the camera shutter or voice still works for foods without a barcode.")
                .font(.system(size: 13))
                .foregroundStyle(Theme.Palette.smoke)
        }
        .padding(20)
        .frame(maxWidth: .infinity, alignment: .leading)
        .background(
            RoundedRectangle(cornerRadius: Theme.Radius.md, style: .continuous)
                .fill(Theme.Palette.inkSurface)
                .overlay(
                    RoundedRectangle(cornerRadius: Theme.Radius.md, style: .continuous)
                        .strokeBorder(Theme.Palette.hairline, lineWidth: 1)
                )
        )
    }

    private func runLookup() {
        let q = trimmedQuery
        guard !q.isEmpty else { return }
        // Only barcode lookups are wired today; the BarcodeAPI client
        // already does the USDA → Open Food Facts chain so we get a real
        // result without writing a parallel resolver here.
        if queryIsCode {
            Task { await lookupCode(q) }
        } else {
            // Free-text name search isn't shipped yet. Be honest about it
            // so the user knows the gap and can fall back to the camera
            // or voice instead.
            lastError = "Free-text name search is coming soon. For now, scan the barcode with the camera or describe it via voice."
            lastResult = nil
        }
    }

    private func lookupCode(_ code: String) async {
        lookingUp = true
        defer { lookingUp = false }
        do {
            let (meal, _) = try await BarcodeAPI.lookup(code: code)
            await MainActor.run {
                lastResult = MealEntry(
                    name: meal.name,
                    detail: meal.detail,
                    calories: meal.kcal,
                    protein: meal.protein_g,
                    carbs: meal.carbs_g,
                    fat: meal.fat_g,
                    loggedAt: .now,
                    slot: MealEntry.Slot(rawValue: meal.slot) ?? .snack,
                    source: .barcode
                )
                lastError = nil
            }
        } catch BarcodeAPI.Error.notFound {
            await MainActor.run {
                lastResult = nil
                lastError = "Couldn't find that code in the food database."
            }
        } catch {
            await MainActor.run {
                lastResult = nil
                lastError = "Lookup failed. \(error.localizedDescription)"
            }
        }
    }
}

#Preview {
    FoodDatabaseSheet()
        .environmentObject(AppModel(
            totals: MockData.today,
            meals: [],
            profile: MockData.profile
        ))
        .preferredColorScheme(.dark)
}
